File: desktop/muzika/dropbox.py
"""The library file, and optionally the music, in Dropbox.

Dropbox has no anonymous mode: it needs a token belonging to the account. The
user generates one in their own app console and pastes it into Settings, so
nothing here ever sees a password and no app secret is baked into the program.
"""

from __future__ import annotations

import json
import logging
from pathlib import Path

import urllib.error
import urllib.request

from . import sync as sync_mod

log = logging.getLogger(__name__)

LIBRARY_PATH = "/Muzika/muzika-library.json"
MUSIC_ROOT = "/Muzika/Music"


class DropboxError(Exception):
    pass


def token() -> str:
    value = (sync_mod.get("dropbox_token") or "").strip()
    if not value:
        raise DropboxError("No Dropbox token — add one in Settings")
    return value


def _request(url: str, *, data: bytes | None, headers: dict) -> bytes:
    request = urllib.request.Request(url, data=data, headers=headers, method="POST")
    try:
        with urllib.request.urlopen(request, timeout=30) as response:
            return response.read()
    except urllib.error.HTTPError as error:
        body = error.read().decode("utf-8", "replace")[:200]
        if error.code == 401:
            raise DropboxError("Dropbox rejected the token") from error
        if error.code == 409:
            raise FileNotFoundError(body) from error
        raise DropboxError(f"Dropbox refused the request ({error.code}): {body}") from error
    except urllib.error.URLError as error:
        raise DropboxError(f"Could not reach Dropbox: {error.reason}") from error


def account_name() -> str:
    """Used by Settings to prove the token actually works."""
    raw = _request(
        "https://api.dropboxapi.com/2/users/get_current_account",
        data=b"null",
        headers={"Authorization": f"Bearer {token()}",
                 "Content-Type": "application/json"},
    )
    data = json.loads(raw or "{}")
    name = (data.get("name") or {}).get("display_name")
    return name or data.get("email") or "Dropbox"


def download(path: str = LIBRARY_PATH) -> bytes | None:
    """None when the file simply is not there yet, which is not an error."""
    try:
        return _request(
            "https://content.dropboxapi.com/2/files/download",
            data=b"",
            headers={"Authorization": f"Bearer {token()}",
                     "Dropbox-API-Arg": json.dumps({"path": path})},
        )
    except FileNotFoundError:
        return None


def upload(content: bytes, path: str = LIBRARY_PATH) -> None:
    argument = json.dumps({"path": path, "mode": "overwrite", "mute": True})
    _request(
        "https://content.dropboxapi.com/2/files/upload",
        data=content,
        headers={"Authorization": f"Bearer {token()}",
                 "Dropbox-API-Arg": argument,
                 "Content-Type": "application/octet-stream"},
    )


# ------------------------------------------------------------------- library

def export_library(store) -> None:
    payload = sync_mod.build_payload(store)
    upload(json.dumps(payload, indent=1, ensure_ascii=False).encode("utf-8"))


def import_library(store) -> dict:
    raw = download()
    if raw is None:
        return {"ok": False, "reason": "no library file in Dropbox yet"}
    try:
        payload = json.loads(raw)
    except ValueError:
        return {"ok": False, "reason": "the Dropbox library file is not readable"}
    if payload.get("format") != sync_mod.FORMAT:
        return {"ok": False, "reason": "not a Muzika library file"}
    return sync_mod.apply_payload(store, payload)


def sync(store) -> dict:
    result = import_library(store)
    export_library(store)
    return result

File: android/app/src/main/java/lt/a777/muzika/data/Dropbox.kt
package lt.a777.muzika.data

import android.util.Log
import lt.a777.muzika.sources.NpeDownloader
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject

/**
 * The library file in Dropbox, over its HTTP API.
 *
 * Dropbox has no anonymous mode: it needs a token that belongs to the account.
 * The user generates one in their own app console and pastes it into settings,
 * so nothing here ever sees a password and no app secret is baked in.
 */
object Dropbox {
    private const val TAG = "Dropbox"
    private const val PATH = "/Muzika/muzika-library.json"

    class Failure(message: String) : Exception(message)

    private fun token(): String = Prefs.dropboxToken.ifEmpty {
        throw Failure("No Dropbox token - add one in Settings")
    }

    /** Null when the file is simply not there yet, which is not an error. */
    fun download(): String? {
        val request = Request.Builder()
            .url("https://content.dropboxapi.com/2/files/download")
            .addHeader("Authorization", "Bearer ${token()}")
            .addHeader("Dropbox-API-Arg", JSONObject().put("path", PATH).toString())
            .post(ByteArray(0).toRequestBody(null))
            .build()
        NpeDownloader.client.newCall(request).execute().use { response ->
            if (response.code == 409) return null          // path/not_found
            if (response.code == 401) throw Failure("Dropbox rejected the token")
            if (!response.isSuccessful) {
                throw Failure("Dropbox download failed (${response.code})")
            }
            return response.body?.string()
        }
    }

    fun upload(content: String) {
        val arg = JSONObject()
            .put("path", PATH)
            .put("mode", "overwrite")
            .put("mute", true)
        val request = Request.Builder()
            .url("https://content.dropboxapi.com/2/files/upload")
            .addHeader("Authorization", "Bearer ${token()}")
            .addHeader("Dropbox-API-Arg", arg.toString())
            .post(content.toByteArray().toRequestBody("application/octet-stream".toMediaType()))
            .build()
        NpeDownloader.client.newCall(request).execute().use { response ->
            if (response.code == 401) throw Failure("Dropbox rejected the token")
            if (!response.isSuccessful) {
                Log.w(TAG, "upload failed ${response.code}")
                throw Failure("Dropbox upload failed (${response.code})")
            }
        }
    }

    /** Used by the settings screen to tell the user the token actually works. */
    fun accountName(): String {
        val request = Request.Builder()
            .url("https://api.dropboxapi.com/2/users/get_current_account")
            .addHeader("Authorization", "Bearer ${token()}")
            .post("null".toRequestBody("application/json".toMediaType()))
            .build()
        NpeDownloader.client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw Failure("Dropbox rejected the token")
            val body = JSONObject(response.body?.string() ?: "{}")
            return body.optJSONObject("name")?.optString("display_name")
                ?.ifEmpty { null } ?: body.optString("email").ifEmpty { "Dropbox" }
        }
    }
}
